hs_labels/models.py

```diff
@@ -0,0 +1,876 @@
+
+    This is typically a programming error and should be entered into
+    issue tracker and resolved.
+
+    *Catching this exception is not recommended.*
+    """
+    def __str__(self):
+        return repr("HS Usage Exception: " + self.message)
+
+    pass
+
+
+####################################
+####################################
+# The labeling system allows
+# a) labeling of resources.
+# b) assigning folders to resources.
+# c) marking public resources to be "of interest"
+####################################
+####################################
+
+
+####################################
+# privileges are numeric codes 1-4
+####################################
+class LabelCodes(object):
+    """
+    Label codes describe the kinds of labels that a user can create for a thing:
+
+        * 1 or LabelCodes.LABEL:
+            a regular label for the object.
+
+        * 2 or LabelCodes.FOLDER:
+            a folder for the object.
+
+        * 3 or LabelCodes.FAVORITE:
+            marks the object as a favorite.
+
+        * 4 or LabelCodes.MINE:
+            marks the object to be included in my resources
+
+        * 5 or LabelCodes.SAVEDLABEL:
+            creates a label that can be used later for labeling a resource
+
+    """
+    LABEL = 1
+    FOLDER = 2
+    FAVORITE = 3
+    MINE = 4
+    SAVEDLABEL = 5
+    LABEL_CHOICES = (
+        (LABEL, 'Label'),
+        (FOLDER, 'Folder'),
+        (FAVORITE, 'Favorite'),
+        (MINE, 'Mine'),
+        (SAVEDLABEL, 'Saved Label')
+    )
+
+####################################
+# user access to resources
+####################################
+
+
+class UserResourceLabels(models.Model):
+    """ Labels of a user for a resource
+
+    This model stores labels of an individual user, like an access
+    control list; There are several kinds of labels documented in LabelCodes.
+    These are similar in implementation but differ in semantics.
+    """
+
+    kind = models.IntegerField(choices=LabelCodes.LABEL_CHOICES,
+                               editable=False,
+                               default=LabelCodes.LABEL)
+    # provenance of label start
+    start = models.DateTimeField(editable=False, auto_now=True)
+
+    # if end is non-null, then the label is no longer valid. Not absolutely sure we need this.
+    # end  = models.DateTimeField(editable=False, null=True, auto_now=False)
+
+    # Utilize an underlying class of User to hold labels for resources.
+    # This creates an implicit query User.ulabels that represents all labels
+    # for a user.
+    ulabels = models.ForeignKey('UserLabels', null=False, editable=False,
+                                related_name='ul2url',         # unused but must be defined and unique
+                                help_text='user assigning a label')
+    # There is a basic design question as to whether the target of a label should be a
+    # Resource proper, or an annotation class that contains mainly methods for the resource.
```

Comment:
We have to balance two things: ease of use and query elegance. In this case, I think there is great utility to making these commands available as part of the User object, in a pattern parallel to that of uaccess. Admittedly, with no data in the intermediate class, this is less "efficient", but in my opinion the UI is easier to understand and the patterns easier to remember.